<div class="announce instapaper_body md" data-path="README.md" id="readme">
 <article class="markdown-body entry-content" itemprop="text">
  <p><a href="https://travis-ci.org/fulcrumgenomics/fgbio" target="_blank"><img src="https://camo.githubusercontent.com/40bbb8b5f7dac6badcef12534c958b0661247105/68747470733a2f2f7472617669732d63692e6f72672f66756c6372756d67656e6f6d6963732f666762696f2e7376673f6272616e63683d6d6173746572" alt="Build Status" data-canonical-src="https://travis-ci.org/fulcrumgenomics/fgbio.svg?branch=master" style="max-width:100%;"></a> <a href="https://codecov.io/gh/fulcrumgenomics/fgbio" target="_blank"><img src="https://camo.githubusercontent.com/9db1e8526c974410d0dafcd608e1117456dfce41/68747470733a2f2f636f6465636f762e696f2f67682f66756c6372756d67656e6f6d6963732f666762696f2f6272616e63682f6d61737465722f67726170682f62616467652e737667" alt="codecov" data-canonical-src="https://codecov.io/gh/fulcrumgenomics/fgbio/branch/master/graph/badge.svg" style="max-width:100%;"></a> <a href="https://www.codacy.com/app/contact_32/fgbio?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=fulcrumgenomics/fgbio&amp;utm_campaign=Badge_Grade" target="_blank"><img src="https://camo.githubusercontent.com/ed7b3a9e9928cdb0fd0293966f0ba23b2c484f24/68747470733a2f2f6170692e636f646163792e636f6d2f70726f6a6563742f62616467652f47726164652f6663346635666538646265333462663738343131343433356232303266616234" alt="Codacy Badge" data-canonical-src="https://api.codacy.com/project/badge/Grade/fc4f5fe8dbe34bf784114435b202fab4" style="max-width:100%;"></a> <a href="https://maven-badges.herokuapp.com/maven-central/com.fulcrumgenomics/fgbio_2.11" target="_blank"><img src="https://camo.githubusercontent.com/5751be0045cacef2a673a48b26640bff159bffff/68747470733a2f2f6d6176656e2d6261646765732e6865726f6b756170702e636f6d2f6d6176656e2d63656e7472616c2f636f6d2e66756c6372756d67656e6f6d6963732f666762696f5f322e31312f62616467652e737667" alt="Maven Central" data-canonical-src="https://maven-badges.herokuapp.com/maven-central/com.fulcrumgenomics/fgbio_2.11/badge.svg" style="max-width:100%;"></a> <a href="https://www.versioneye.com/user/projects/57a1584c3d8eb6002dc1e812#dialog_dependency_badge" target="_blank"><img src="https://camo.githubusercontent.com/79efac98810ecb10ae3fb457d3be33c2c20b9ce3/68747470733a2f2f7777772e76657273696f6e6579652e636f6d2f757365722f70726f6a656374732f3537613135383463336438656236303032646331653831322f62616467652e737667" alt="Dependency Status" data-canonical-src="https://www.versioneye.com/user/projects/57a1584c3d8eb6002dc1e812/badge.svg" style="max-width:100%;"></a> <a href="https://github.com/fulcrumgenomics/fgbio/blob/master/LICENSE" target="_blank"><img src="https://camo.githubusercontent.com/e0d5267d60ee425acfe1a1f2d6e6d92a465dcd8f/687474703a2f2f696d672e736869656c64732e696f2f62616467652f6c6963656e73652d4d49542d626c75652e737667" alt="License" data-canonical-src="http://img.shields.io/badge/license-MIT-blue.svg" style="max-width:100%;"></a> <a href="http://www.scala-lang.org/" target="_blank"><img src="https://camo.githubusercontent.com/37a3e182c9195167ec4832444e6dd9fb0e1cd1b8/687474703a2f2f696d672e736869656c64732e696f2f62616467652f6c616e67756167652d7363616c612d627269676874677265656e2e737667" alt="Language" data-canonical-src="http://img.shields.io/badge/language-scala-brightgreen.svg" style="max-width:100%;"></a></p> 
  <h1><a id="user-content-fgbio" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#fgbio" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>fgbio</h1> 
  <p>A set of tools to analyze genomic data.</p> 
  <ul> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#goals" target="_blank">Goals</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#building" target="_blank">Building</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#command-line" target="_blank">Command line</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#include-fgbio-in-your-project" target="_blank">Include fgbio in your project</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#overview" target="_blank">Overview</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#list-of-tools" target="_blank">List of tools</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#contributing" target="_blank">Contributing</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#authors" target="_blank">Authors</a></li> 
   <li><a href="https://github.com/fulcrumgenomics/fgbio#license" target="_blank">License</a></li> 
  </ul> 
  <h1><a id="user-content-goals" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#goals" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Goals</h1> 
  <p>There are many toolkits available for analyzing genomic data; fgbio does not aim to be all things to all people but is specifically focused on providing:</p> 
  <ul> 
   <li>Robust, well-tested tools.</li> 
   <li>An easy to use command-line.</li> 
   <li>Documentation for each tool.</li> 
   <li>Tools not found anywhere else.</li> 
   <li>Open source development for the benefit of the community and our clients.</li> 
  </ul> 
  <h2><a id="user-content-building" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#building" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Building</h2> 
  <h3><a id="user-content-cloning-the-repository" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#cloning-the-repository" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Cloning the Repository</h3> 
  <p><a href="https://git-lfs.github.com/" target="_blank">Git LFS</a> is used to store large files used in testing fgbio. In order to compile and run tests it is necessary to <a href="https://git-lfs.github.com/" target="_blank">install git lfs</a>. To retrieve the large files either:</p> 
  <ol> 
   <li>Clone the repository <em>after</em> installing git lfs, or</li> 
   <li>In a previously cloned repository run <code>git lfs pull</code> once</li> 
  </ol> 
  <p>After initial setup regular git commands (e.g. <code>pull</code>, <code>fetch</code>, <code>push</code>) will also operate on large files and no special handling is needed.</p> 
  <p>To clone the repository: <code>git clone https://github.com/fulcrumgenomics/fgbio.git</code></p> 
  <h3><a id="user-content-running-the-build" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#running-the-build" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Running the build</h3> 
  <p>fgbio is built using <a href="http://www.scala-sbt.org/" target="_blank">sbt</a>.</p> 
  <p>Use <code>sbt assembly</code> to build an executable jar in <code>target/scala-2.11/</code>.<br> Tests may be run with <code>sbt test</code>. Java SE 8 is required.</p> 
  <h2><a id="user-content-command-line" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#command-line" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Command line</h2> 
  <p><code>java -jar target/scala-2.11/fgbio-0.1.3-SNAPSHOT.jar</code> to see the commands supported. Use <code>java -jar target/scala-2.11/fgbio-0.1.3-SNAPSHOT.jar &lt;command&gt;</code> to see the help message for a particular command.</p> 
  <h2><a id="user-content-include-fgbio-in-your-project" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#include-fgbio-in-your-project" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Include fgbio in your project</h2> 
  <p>You can include <code>fgbio</code> in your project:</p> 
  <pre><code>"com.fulcrumgenomics" %% "fgbio" % "0.1.3-SNAPSHOT"
</code></pre> 
  <h2><a id="user-content-overview" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#overview" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Overview</h2> 
  <p>Fgbio is a command line tool to perform bioinformatic genomic data analysis. The collection of tools within <code>fgbio</code> are used by our customers for both ad-hoc data analysis and within their production pipelines. These tools typically operate on read-level data (ex. FASTQ, SAM, or BAM) or variant-level data (ex. VCF or BCF). They range from simple tools to filter reads in a BAM file, to tools to compute consensus reads from reads with the same molecular index/tag. See the <a href="https://github.com/fulcrumgenomics/fgbio#list-of-tools" target="_blank">list of tools</a> for more detail on the tools</p> 
  <h2><a id="user-content-list-of-tools" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#list-of-tools" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>List of tools</h2> 
  <p>Below we highlight a few tools that you may find useful. Please see the help message for a full list of tools available. In no particular order ...</p> 
  <ul> 
   <li>Tools to work with unique molecular tags/indexes (Umis). 
    <ul> 
     <li>Annotating/Extract Umis from read-level data: <code>AnnotateBamWithUmis</code> and <code>ExtractUmisFromBam</code>.</li> 
     <li>Tools to manipulate read-level data containing Umis: <code>CallMolecularConsensusReads</code> and <code>GroupReadsByUmi</code></li> 
    </ul> </li> 
   <li>Tools to manipulate read-level data: 
    <ul> 
     <li>Filter read-level data: <code>FilterBam</code>.</li> 
     <li>Randomize the order of read-level data: <code>RandomizeBam</code>.</li> 
     <li>Update read-level metadata: <code>SetMateInformation</code> and <code>UpdateReadGroups</code>.</li> 
    </ul> </li> 
   <li>Miscellaneous tools: 
    <ul> 
     <li>Pick molecular indices (ex. sample barcodes, or molecular indexes): <code>PickIlluminaIndices</code>.</li> 
     <li>Convert the output of HAPCUT (a tool for phasing variants): <code>HapCutToVcf</code>.</li> 
     <li>Find technical or synthetic sequences in read-level data: <code>FindTechnicalReads</code>.</li> 
     <li>Assess phased variant calls: <code>AssessPhasing</code>.</li> 
    </ul> </li> 
  </ul> 
  <h2><a id="user-content-contributing" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#contributing" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Contributing</h2> 
  <p>Contributions are welcome and encouraged. We will do our best to provide an initial response to any pull request or issue within one-week. For urgent matters, please contact us directly.</p> 
  <h2><a id="user-content-authors" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#authors" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Authors</h2> 
  <ul> 
   <li><a href="https://github.com/tfenne" target="_blank">Tim Fennell</a> (maintainer)</li> 
   <li><a href="https://github.com/nh13" target="_blank">Nils Homer</a> (maintainer)</li> 
  </ul> 
  <h2><a id="user-content-license" class="anchor" href="https://github.com/fulcrumgenomics/fgbio#license" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>License</h2> 
  <p><code>fgbio</code> is open source software released under the <a href="https://github.com/fulcrumgenomics/fgbio/blob/master/LICENSE" target="_blank">MIT License</a>.</p> 
 </article>
</div>