<div class="announce instapaper_body md" data-path="README.md" id="readme">
 <article class="markdown-body entry-content" itemprop="text">
  <p><a href="https://github.com/tmoerman/vcf-comp/blob/master/img/venn.png" target="_blank"><img src="https://github.com/tmoerman/vcf-comp/raw/master/img/venn.png" alt="venn" style="max-width:100%;"></a></p> 
  <h1><a href="https://github.com/tmoerman/vcf-comp#vcf-comp" aria-hidden="true" class="anchor" id="user-content-vcf-comp" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>VCF-comp</h1> 
  <p>VCF-comp is a <a href="http://www.scala-lang.org/" target="_blank">Scala</a> library for pairwise comparison of annotated <a href="http://samtools.github.io/hts-specs/VCFv4.2.pdf" target="_blank">VCF</a> files. Uses <a href="http://spark.apache.org/" target="_blank">Apache Spark</a>, <a href="https://github.com/bigdatagenomics/adam" target="_blank">ADAM</a> and <a href="https://github.com/tmoerman/adam-fx" target="_blank">adam-fx</a>.</p> 
  <p>VCF-comp is intended for performing VCF analyses using the Scala programming language within a <a href="https://github.com/andypetrella/spark-notebook" target="_blank">Spark-notebook</a> environment.</p> 
  <p>VCF-comp is open source software, available on both <a href="https://github.com/tmoerman/vcf-comp" target="_blank">Github</a> and <a href="https://bitbucket.org/vda-lab/vcf-comp" target="_blank">BitBucket</a>.</p> 
  <p>VCF-comp artifacts are published to <a href="https://bintray.com/tmoerman/maven/vcf-comp" target="_blank">Bintray</a>. Latest version: <code>0.4.0</code></p> 
  <p>--</p> 
  <ul> 
   <li><a href="https://github.com/tmoerman/vcf-comp#vcf-comp" target="_blank">VCF-comp</a> 
    <ul> 
     <li><a href="https://github.com/tmoerman/vcf-comp#change-notes" target="_blank">CHANGE NOTES</a></li> 
     <li><a href="https://github.com/tmoerman/vcf-comp#getting-started" target="_blank">GETTING STARTED</a> 
      <ul> 
       <li><a href="https://github.com/tmoerman/vcf-comp#using-the-docker-image" target="_blank">Using the Docker image</a></li> 
       <li><a href="https://github.com/tmoerman/vcf-comp#manual-setup" target="_blank">Manual setup</a> 
        <ul> 
         <li><a href="https://github.com/tmoerman/vcf-comp#remote-artifact-repository" target="_blank">Remote artifact repository</a></li> 
         <li><a href="https://github.com/tmoerman/vcf-comp#library-dependencies" target="_blank">Library dependencies</a></li> 
         <li><a href="https://github.com/tmoerman/vcf-comp#initialize-the-sparkcontext" target="_blank">Initialize the SparkContext</a></li> 
         <li><a href="https://github.com/tmoerman/vcf-comp#import-vcf-comp-functionality" target="_blank">Import VCF-comp functionality</a></li> 
         <li><a href="https://github.com/tmoerman/vcf-comp#test-the-setup" target="_blank">Test the setup</a></li> 
        </ul> </li> 
      </ul> </li> 
     <li><a href="https://github.com/tmoerman/vcf-comp#usage" target="_blank">USAGE</a> 
      <ul> 
       <li><a href="https://github.com/tmoerman/vcf-comp#starting-a-qc-comparison" target="_blank">Starting a QC comparison</a></li> 
       <li><a href="https://github.com/tmoerman/vcf-comp#starting-a-snp-comparison" target="_blank">Starting a SNP comparison</a></li> 
      </ul> </li> 
     <li><a href="https://github.com/tmoerman/vcf-comp#concordance-categories" target="_blank">CONCORDANCE CATEGORIES</a></li> 
     <li><a href="https://github.com/tmoerman/vcf-comp#overview-of-analyses" target="_blank">OVERVIEW OF ANALYSES</a> 
      <ul> 
       <li><a href="https://github.com/tmoerman/vcf-comp#qc" target="_blank">QC</a> 
        <ul> 
         <li><a href="https://github.com/tmoerman/vcf-comp#qc-analyses" target="_blank">QC analyses</a></li> 
         <li><a href="https://github.com/tmoerman/vcf-comp#qc-analysis-building-blocks" target="_blank">QC analysis building blocks</a></li> 
        </ul> </li> 
       <li><a href="https://github.com/tmoerman/vcf-comp#snp-comparison" target="_blank">SNP comparison</a> 
        <ul> 
         <li><a href="https://github.com/tmoerman/vcf-comp#snp-comparison-analyses" target="_blank">SNP comparison analyses</a></li> 
         <li><a href="https://github.com/tmoerman/vcf-comp#snp-comparison-building-blocks" target="_blank">SNP comparison building blocks</a></li> 
        </ul> </li> 
       <li><a href="https://github.com/tmoerman/vcf-comp#visualizations" target="_blank">Visualizations</a></li> 
      </ul> </li> 
     <li><a href="https://github.com/tmoerman/vcf-comp#how-it-works" target="_blank">HOW IT WORKS</a> 
      <ul> 
       <li><a href="https://github.com/tmoerman/vcf-comp#pimp-my-library" target="_blank">Pimp my library</a></li> 
       <li><a href="https://github.com/tmoerman/vcf-comp#dimplejs" target="_blank">Dimple.js</a></li> 
      </ul> </li> 
    </ul> </li> 
  </ul> 
  <h2><a href="https://github.com/tmoerman/vcf-comp#change-notes" aria-hidden="true" class="anchor" id="user-content-change-notes" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>CHANGE NOTES</h2> 
  <ul> 
   <li><code>0.4.0</code> Added filtering functions: 
    <ul> 
     <li>filtering by gene name(s) with respect to .bed and .gtf files</li> 
     <li>filtering by contig</li> 
    </ul> </li> 
   <li><code>0.3.0</code> Stable version</li> 
  </ul> 
  <h2><a href="https://github.com/tmoerman/vcf-comp#getting-started" aria-hidden="true" class="anchor" id="user-content-getting-started" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>GETTING STARTED</h2> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#using-the-docker-image" aria-hidden="true" class="anchor" id="user-content-using-the-docker-image" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Using the Docker image</h3> 
  <p><code>TODO</code></p> 
  <p><em>(A Docker image containing Spark Notebook and example notebooks with VCF-comp analyses will be available in the near future)</em></p> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#manual-setup" aria-hidden="true" class="anchor" id="user-content-manual-setup" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Manual setup</h3> 
  <p>We will assume you have a <a href="https://github.com/andypetrella/spark-notebook" target="_blank">Spark-notebook</a> instance available. If not, have a look at the <a href="https://github.com/andypetrella/spark-notebook#using-a-release" target="_blank">launch instructions</a>.</p> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#remote-artifact-repository" aria-hidden="true" class="anchor" id="user-content-remote-artifact-repository" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Remote artifact repository</h4> 
  <p>First, we specify the remote Maven <a href="https://bintray.com/tmoerman/maven/vcf-comp" target="_blank">repository</a> on Bintray from which the VCF-comp library artifact are available.</p> 
  <pre><code>:remote-repo bintray-tmoerman % default % http://dl.bintray.com/tmoerman/maven % maven
</code></pre> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#library-dependencies" aria-hidden="true" class="anchor" id="user-content-library-dependencies" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Library dependencies</h4> 
  <p>Next, we specify the library dependencies. We need both the VCF-comp library and the BDGenomics Adam library, but without Hadoop and Spark dependencies, because these are already provided automatically in the Spark Notebook environment.</p> 
  <pre><code>:dp org.tmoerman % vcf-comp_2.10 % 0.4.0
:dp org.bdgenomics.adam % adam-core % 0.17.1
- org.apache.hadoop % hadoop-client %   _
- org.apache.spark  %     _         %   _
- org.scala-lang    %     _         %   _
- org.scoverage     %     _         %   _
- joda-time         %     _         %   _
</code></pre> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#initialize-the-sparkcontext" aria-hidden="true" class="anchor" id="user-content-initialize-the-sparkcontext" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Initialize the SparkContext</h4> 
  <p>We are now ready to configure the running <code>SparkContext</code> in the notebook. VCF-comp uses Adam's rich domain classes, therefore we need to define Kryo as the Spark serializer and the Adam-FX Kryo registrator. Don't worry too much about this, just execute the following snippet in a Spark Notebook cell.</p> 
  <pre><code>reset(lastChanges = _.set("spark.app.name", "My VCF analysis")
                     .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
                     .set("spark.kryo.registrator", "org.tmoerman.adam.fx.serialization.AdamFxKryoRegistrator")
                     .set("spark.kryoserializer.buffer", "32m")
                     .set("spark.kryoserializer.buffer.max value", "128")
                     .set("spark.kryo.referenceTracking", "true"))
</code></pre> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#import-vcf-comp-functionality" aria-hidden="true" class="anchor" id="user-content-import-vcf-comp-functionality" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Import VCF-comp functionality</h4> 
  <p>Almost there! The final setup step is to import the necessary VCF-comp classes and functions.</p> 
  <pre><code>import scala.language.implicitConversions

import org.tmoerman.vcf.comp.VcfComparisonContext._
import org.tmoerman.vcf.comp.core.Model._
import org.tmoerman.vcf.comp.viz.SparkNoteBookDimpleGraphs._

import org.tmoerman.adam.fx.avro.AnnotatedGenotype

implicit def toDimpleChart(chart: DimpleChart) = chart match {
  case DimpleChart(data, js, s) =&gt; DiyChart(data, js, maxPoints = chart.maxPoints, sizes = s)
}
</code></pre> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#test-the-setup" aria-hidden="true" class="anchor" id="user-content-test-the-setup" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Test the setup</h4> 
  <p>We can now test the setup of the library by executing the <code>.help</code> function on the SparkContext variable (<code>sc</code> or <code>sparkContext</code>).</p> 
  <pre><code>sc.help
</code></pre> 
  <p>This method should return a list of methods we can invoke on the SparkContext:</p> 
  <pre><code>res9: String =
- getMetaFields
- startQcComparison
- startSnpComparison
</code></pre> 
  <p>This is VCF-comp's so-called <em>discoverable API</em> in action. We discuss this concept in more detail later.</p> 
  <p>Well done! We are now ready to perform an actual pairwise VCF comparison analysis! Read on to find out how.</p> 
  <h2><a href="https://github.com/tmoerman/vcf-comp#usage" aria-hidden="true" class="anchor" id="user-content-usage" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>USAGE</h2> 
  <p>Some proficiency in <a href="http://www.scala-lang.org/" target="_blank">Scala</a> is expected. Don't worry, the level of Scala programming required to effectively use this library is quite basic. If you have some experience with Java, Python, C++ or C#, most of the code in this tutorial will feel very familiar.</p> 
  <p>The motivation for adopting a programmatic Notebook approach in this tool is because this is essentially a <em>data science</em> tool. Data science tools ideally show exactly how an analysis has been performed, by making the individual steps explicit as code. Although the threshold to get started is higher than a point-and-click interface, this disadvantage is outweighed in the long run by the benefits of the reproducible nature of a notebook approach.</p> 
  <p>VCF-comp focuses on pairwise comparison of VCF files, so let's get two interesting files ready: <code>tumor.vcf</code> and <code>normal.vcf</code>. In this example, we read these files from <a href="https://hadoop.apache.org/docs/r1.2.1/hdfs_design.html" target="_blank">HDFS</a>:</p> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">val</span> <span class="pl-en">workDir</span> <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>hdfs://bla:54310/my-data/VCF/<span class="pl-pds">"</span></span>

<span class="pl-k">val</span> <span class="pl-en">tumor</span>  <span class="pl-k">=</span> workDir <span class="pl-k">+</span> <span class="pl-s"><span class="pl-pds">"</span>tumor.vcf<span class="pl-pds">"</span></span>
<span class="pl-k">val</span> <span class="pl-en">normal</span> <span class="pl-k">=</span> workDir <span class="pl-k">+</span> <span class="pl-s"><span class="pl-pds">"</span>normal.vcf<span class="pl-pds">"</span></span></pre>
  </div> 
  <p>We can now start two types of VCF comparison: a quality control (QC) comparison and a SNP comparison. Let's start with a QC comparison.</p> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#starting-a-qc-comparison" aria-hidden="true" class="anchor" id="user-content-starting-a-qc-comparison" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Starting a QC comparison</h3> 
  <p>To start a QC comparison, we invoke the appropriate method on the SparkContext instance <code>sc</code>. Remember, as previously mentioned, we can always invoke the <code>.help</code> method on different objects in the analysis, to give us a list of methods available at that point. Let's do that one more time:</p> 
  <div class="highlight highlight-source-scala">
   <pre>sc.help</pre>
  </div> 
  <pre><code>res9: String =
- getMetaFields
- startQcComparison
- startSnpComparison
</code></pre> 
  <p>The method <code>startQcComparison</code> is the one we need, let's invoke it while setting some parameters:</p> 
  <pre><code>val qcParams = new ComparisonParams(labels = ("TUMOR", "NORMAL")) // parameters

val qcComparison = sc.startQcComparison(tumor, normal, qcParams) // a Spark RDD
                     .cache()                                    // cache the RDD
</code></pre> 
  <p>The <code>ComparisonParams</code> object defines labels for our VCF files. The result of invoking <code>startQcComparison</code> is a Spark <a href="http://spark.apache.org/docs/latest/programming-guide.html#resilient-distributed-datasets-rdds" target="_blank">RDD</a>, which we cache in memory because it is an intermediate result we will use in future computations. We store this RDD in a value named <code>qcComparison</code>.</p> 
  <p>Let's once again use the <code>.help</code> method to discover how we can proceed. This time we don't invoke it on the SparkContext <code>sc</code>, but on the <code>qcComparison</code> value.</p> 
  <div class="highlight highlight-source-scala">
   <pre>qcComparison.help</pre>
  </div> 
  <pre><code>res11: String =
- countByProjection
- countByTraversableProjection
- indelLengthDistribution
- snpCountByContig
- snpQualityDistribution
- snpReadDepthDistribution
- variantTypeCount
</code></pre> 
  <p>Looks like we can do some interesting analyses! Let's have a look at the number of SNPs we can find in the different contigs, so we choose <code>snpCountByContig</code>.</p> 
  <div class="highlight highlight-source-scala">
   <pre>qcComparison.snpCountByContig</pre>
  </div> 
  <p>Oops, now we get some gobbledigook:</p> 
  <pre><code>res16: Iterable[org.tmoerman.vcf.comp.core.Model.QcProjectionCount[String]] = List(QcProjectionCount(NORMAL,9,1361),
QcProjectionCount(NORMAL,1,3304), QcProjectionCount(TUMOR,20,844), QcProjectionCount(NORMAL,12,1564),
QcProjectionCount(TUMOR,19,2607), QcProjectionCount(TUMOR,21,476), QcProjectionCount(TUMOR,1,3376),
QcProjectionCount(TUMOR,13,547), QcProjectionCount(TUMOR,Y,12), QcProjectionCount(NORMAL,11,2300),
QcProjectionCount(TUMOR,22,733), QcProjectionCount(NORMAL,2,2197), QcProjectionCount(TUMOR,12,1520),
QcProjectionCount(NORMAL,7,1522), QcProjectionCount(NORMAL,16,1387), QcProjectionCount(NORMAL,18,54...
</code></pre> 
  <p>That doesn't look right, we'd prefer to see some graphical output. Let's consult the <code>.help</code> function once more, this time we invoke it on the result of the <code>qcComparison.snpCountByContig</code> calculation:</p> 
  <div class="highlight highlight-source-scala">
   <pre>qcComparison.snpCountByContig.help</pre>
  </div> 
  <pre><code>res17: String =
- groupedBarChart
- lineChart
</code></pre> 
  <p>Okay, a histogram is probably the most sensible chart for a SNP count by contig, so let's choose that one:</p> 
  <div class="highlight highlight-source-scala">
   <pre>qcComparison.snpCountByContig
            .groupedBarChart(x_title <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>SNP count by contig<span class="pl-pds">"</span></span>,
                             x_order <span class="pl-k">=</span> <span class="pl-c1">true</span>)</pre>
  </div> 
  <p>We also specified some overriding attributes of the grouped bar chart. The result is an object that is automagically turned into a chart by the notebook.</p> 
  <p><a href="https://github.com/tmoerman/vcf-comp/blob/master/img/snpCount.png" target="_blank"><img src="https://github.com/tmoerman/vcf-comp/raw/master/img/snpCount.png" alt="chart" style="max-width:100%;"></a></p> 
  <p>Nice!</p> 
  <p>This sequence of steps illustrates the primary usage pattern of the VCF-comp library. See section <a href="https://github.com/tmoerman/vcf-comp#" target="_blank">QC analyses</a> for an overview of QC functionality.</p> 
  <p>Let's now apply this for a SNP comparison.</p> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#starting-a-snp-comparison" aria-hidden="true" class="anchor" id="user-content-starting-a-snp-comparison" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Starting a SNP comparison</h3> 
  <p>Analogously, we launch a SNP comparison.</p> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">val</span> <span class="pl-en">snpParams</span> <span class="pl-k">=</span> <span class="pl-k">new</span> <span class="pl-en">ComparisonParams</span>(labels <span class="pl-k">=</span> (<span class="pl-s"><span class="pl-pds">"</span>TUMOR<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>NORMAL<span class="pl-pds">"</span></span>))

<span class="pl-k">val</span> <span class="pl-en">snpComparison</span> <span class="pl-k">=</span> sc.startSnpComparison(tumor, normal, snpParams) <span class="pl-c"><span class="pl-c">//</span> a Spark RDD</span>
                      .cache()                                      <span class="pl-c"><span class="pl-c">//</span> cache the RDD		</span></pre>
  </div> 
  <p>In the same spirit as the QC example, we create a Spark RDD instance by invoking the <code>startSnpComparison</code> function on the SparkContext <code>sc</code>. Analogously, we can inspect the list of available functions by invoking the <code>.help</code> function on the <code>snpComparison</code> value.</p> 
  <div class="highlight highlight-source-scala">
   <pre>snpComparison.help</pre>
  </div> 
  <pre><code>res17: String =
- alleleFrequencyDistribution
- allelesSwitchCount
- baseChangeCount
- baseChangePatternCount
- baseChangeTypeCount
- categoryCount
- clinvarRatio
- commonSnpRatio
- countByProjection
- countBySwitch
- countByTraversableProjection
- functionalAnnotationCount
- functionalImpactCount
- qualityDistribution
- readDepthDistribution
- synonymousRatio
- transcriptBiotypeCount
- viewOnly
- zygosityCount
- zygositySwitchCount
</code></pre> 
  <p>Let's choose the <code>readDepthDistribution</code> analysis, and inspect which visualizations are available. Note that not all visualization make sense for a "distribution" type analysis.</p> 
  <div class="highlight highlight-source-scala">
   <pre>snpComparison.readDepthDistribution().help</pre>
  </div> 
  <pre><code>res21: String =
- groupedBarChart
- lineChart
- lollipopPieChart
- percentageBarChart
- stackedBarChart
- table
</code></pre> 
  <p>In this case, a line chart makes most sense, so let's try that.</p> 
  <div class="highlight highlight-source-scala">
   <pre>snpComparison.readDepthDistribution(step <span class="pl-k">=</span> <span class="pl-c1">5</span>)
             .lineChart(width <span class="pl-k">=</span> <span class="pl-c1">700</span>,
                        x_title <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>read depth distribution<span class="pl-pds">"</span></span>)</pre>
  </div> 
  <p><a href="https://github.com/tmoerman/vcf-comp/blob/master/img/readDepth.png" target="_blank"><img src="https://github.com/tmoerman/vcf-comp/raw/master/img/readDepth.png" alt="RD" style="max-width:100%;"></a></p> 
  <p>Sweet!</p> 
  <p>Notice that again, we have overridden some of the default arguments in both the <code>readDepthDistribution</code> as well as the <code>lineChart</code> methods. Check the documentation for more information about which arguments are available on the analysis methods.</p> 
  <p>This concludes two examples of how to launch a QC comparison and a SNP comparison. The next section deals with the core concept of the SNP comparison functionality: the 5 concordance categories.</p> 
  <h2><a href="https://github.com/tmoerman/vcf-comp#concordance-categories" aria-hidden="true" class="anchor" id="user-content-concordance-categories" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>CONCORDANCE CATEGORIES</h2> 
  <p>The heart of VCF-comp is an algorithm that matches variants per position by concordance. By concordance, we mean a degree to which both files agree about the variants or their genotypes on that position, according to a configurable matching criterion. VCF-comp defines 5 categories:</p> 
  <table> 
   <thead> 
    <tr> 
     <th>name</th> 
     <th>meaning</th> 
    </tr> 
   </thead> 
   <tbody> 
    <tr> 
     <td><code>A-unique</code></td> 
     <td>File A has a variant on this position, file B does not</td> 
    </tr> 
    <tr> 
     <td><code>B-unique</code></td> 
     <td>File B has a variant on this position, file A does not</td> 
    </tr> 
    <tr> 
     <td><code>Concordant</code></td> 
     <td>Both file A and B have the same variant on this position with respect <br> to the matching criterion. A concordant variant is counted once.</td> 
    </tr> 
    <tr> 
     <td><code>A-discordant</code></td> 
     <td>File A and B have a variant on this position, but do not agree with <br> respect to the matching criterion. This variant is the one from file A.</td> 
    </tr> 
    <tr> 
     <td><code>B-discordant</code></td> 
     <td>File A and B have a variant on this position, but do not agree with <br> respect to the matching criterion. This variant is the one from file B.</td> 
    </tr>
   </tbody>
  </table> 
  <p>The default matching criterion is: matching <strong>genotype alleles</strong> in both files A and B.</p> 
  <p>In our opinion, this is the most sensible default behaviour. However, a researcher might want to define a different matching criterion. This is possible. The <code>ComparisonParams</code> class defines the default behaviour, but this is overridable, as illustrated in following example:</p> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">val</span> <span class="pl-en">snpParamsAA</span> <span class="pl-k">=</span> <span class="pl-k">new</span> <span class="pl-en">ComparisonParams</span>(
  labels <span class="pl-k">=</span> (<span class="pl-s"><span class="pl-pds">"</span>TUMOR<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>NORMAL<span class="pl-pds">"</span></span>),
  matchFunction <span class="pl-k">=</span> (<span class="pl-v">gt</span>: <span class="pl-en">AnnotatedGenotype</span>) <span class="pl-k">=&gt;</span> gt.getGenotype.getVariant.getAlternateAllele) <span class="pl-c"><span class="pl-c">//</span> overridden match function</span>

<span class="pl-k">val</span> <span class="pl-en">snpComparisonAA</span> <span class="pl-k">=</span> sc.startSnpComparison(tumor, normal, snpParamsAA)
                        .cache()</pre>
  </div> 
  <p>The <code>matchFunction</code> is the function used in the matching algorithm to determine concordance. The matchFunction can return anything, or in Scala jargon: the function has the <code>Any</code> return type. If file A and B have a variant on a position, this function is applied on the corresponding genotypes and the results are compared for equality. If the results are equal we have a concordant variant, if they are not equal we have two discordant variants.</p> 
  <p>Here we have overridden it with a function that takes an <code>AnnotatedGenotype</code> and returns the alternate allele of the variant associated with the genotype. This is a less strict matching criterion than the default one, and could be useful for a particular analysis.</p> 
  <p>Note that overriding the <code>matchFunction</code> only affects the balance between concordant and discordant variants. The unique variants remain stable.</p> 
  <h2><a href="https://github.com/tmoerman/vcf-comp#overview-of-analyses" aria-hidden="true" class="anchor" id="user-content-overview-of-analyses" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>OVERVIEW OF ANALYSES</h2> 
  <p>This section contains an overview of all available analyses. Additionally, the <a href="https://en.wikipedia.org/wiki/Higher-order_function" target="_blank">higher-order</a> functions that act as building blocks for the available analyses and ad-hoc analyses are discussed. Using the building block functions requires a more advanced mastery of the Scala language.</p> 
  <p><code>TODO</code> <em>(Scala docs will be hosted on a dedicated site in the near future.)</em></p> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#qc" aria-hidden="true" class="anchor" id="user-content-qc" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>QC</h3> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#qc-analyses" aria-hidden="true" class="anchor" id="user-content-qc-analyses" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>QC analyses</h4> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">variantTypeCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">snpCountByContig</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre>snpReadDepthDistribution(<span class="pl-v">step</span>: <span class="pl-k">Int</span> <span class="pl-k">=</span> <span class="pl-en">DEFAULT_READ_DEPTH_STEP</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">snpQualityDistribution</span>(<span class="pl-v">step</span>: <span class="pl-k">Double</span> <span class="pl-k">=</span> <span class="pl-en">DEFAULT_QUALITY_STEP</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">indelLengthDistribution</span></pre>
  </div> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#qc-analysis-building-blocks" aria-hidden="true" class="anchor" id="user-content-qc-analysis-building-blocks" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>QC analysis building blocks</h4> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">countByProjection</span>[<span class="pl-en">P</span><span class="pl-k">:</span> <span class="pl-en">ClassTag</span>](<span class="pl-v">projection</span>: <span class="pl-en">VariantContext</span> <span class="pl-k">=&gt;</span> <span class="pl-en">P</span>)<span class="pl-k">:</span> <span class="pl-en">Iterable</span>[<span class="pl-en">QcProjectionCount</span>[<span class="pl-en">P</span>]]</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">countByTraversableProjection</span>[<span class="pl-en">P</span><span class="pl-k">:</span> <span class="pl-en">ClassTag</span>](<span class="pl-v">projection</span>: <span class="pl-en">VariantContext</span> <span class="pl-k">=&gt;</span> <span class="pl-en">Traversable</span>[<span class="pl-en">P</span>])<span class="pl-k">:</span> <span class="pl-en">Iterable</span>[<span class="pl-en">QcProjectionCount</span>[<span class="pl-en">P</span>]]</pre>
  </div> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#snp-comparison" aria-hidden="true" class="anchor" id="user-content-snp-comparison" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>SNP comparison</h3> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#snp-comparison-analyses" aria-hidden="true" class="anchor" id="user-content-snp-comparison-analyses" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>SNP comparison analyses</h4> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">occurrences</span> vararg. Accepts one or more Strings from "unique", "concordant", "discordant".</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the RDD, filtered on the specified occurrences.</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">viewOnly</span>(<span class="pl-v">occurrences</span>: <span class="pl-k">String</span><span class="pl-k">*</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by concordance category.</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">categoryCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by base change per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Base change is a String: "ref-&gt;alt" where ref and alt are the variant alleles, e.g. "A-&gt;T", "G-&gt;C", etc...</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">baseChangeCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by base change pattern per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Base change pattern is a String: "ref:alt", analogous to base change, but without taking into account</span>
<span class="pl-c">  *         the order of ref to alt, e.g. "A:T", "C:G", etc...</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">baseChangePatternCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by base change type per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Base change types are "Ti" (Transition) and "Tv" (Transversion).</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">baseChangeTypeCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by zygosity per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Zygosity values are: "HOM_REF", "HET", "HOM_ALT", "NO_CALL".</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">zygosityCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by functional impact per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Functional impact is a scale value provided by SnpEff: "HIGH", "MODERATE", "LOW" and "MODIFIER".</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Cfr. SnpEff http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">functionalImpactCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by functional annotation per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Functional annotation is a SnpEff annotation, including: "synonymous_variant", "missense_variant",</span>
<span class="pl-c">  *         "stop_gained", "start_lost".</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Assumes annotation with SnpEff http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">functionalAnnotationCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns SNP count by transcript biotype per concordance category.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Transcript biotype is a SnpEff annotation: including: "protein_coding", "retained_intron"</span>
<span class="pl-c">  *         "nonsense_mediated_decay".</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Assumes annotation with SnpEff http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">transcriptBiotypeCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">label</span> (optional). Maps the Boolean to a descriptive label.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the ratio of SNPs with a Clinvar annotation vs. SNPs without.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Assumes VCF annotation with SnpSift http://snpeff.sourceforge.net/SnpSift.html</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">clinvarRatio</span>(<span class="pl-v">label</span>: <span class="pl-k">Boolean</span> <span class="pl-k">=&gt;</span> <span class="pl-k">String</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">label</span> (optional). Maps the Boolean to a descriptive label.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the ratio of SNPs with a DbSNP annotation vs. SNPs without.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Assumes VCF annotation with SnpSift http://snpeff.sourceforge.net/SnpSift.html</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">commonSnpRatio</span>(<span class="pl-v">label</span>: <span class="pl-k">Boolean</span> <span class="pl-k">=&gt;</span> <span class="pl-k">String</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">label</span> (optional). Maps the boolean to a descriptive label.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the ratio of SNPs with "synonymous_variant" vs. "missense_variant" annotation. If a SNP has</span>
<span class="pl-c">  *         neither annotation, it is not taken into account.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  *         Assumes annotation with SnpEff http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">synonymousRatio</span>(<span class="pl-v">label</span>: <span class="pl-k">Boolean</span> <span class="pl-k">=&gt;</span> <span class="pl-k">String</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">step</span> (optional). Size of the step interval for binning the read depth values.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the distribution of SNPs by read depth.</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">readDepthDistribution</span>(<span class="pl-v">step</span>: <span class="pl-k">Int</span> <span class="pl-k">=</span> <span class="pl-en">DEFAULT_READ_DEPTH_STEP</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">step</span> (optional). Size of the step interval for binning the quality values.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the distribution of SNPs by quality.</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">qualityDistribution</span>(<span class="pl-v">step</span>: <span class="pl-k">Double</span> <span class="pl-k">=</span> <span class="pl-en">DEFAULT_QUALITY_STEP</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-c"><span class="pl-c">/**</span></span>
<span class="pl-c">  * <span class="pl-k">@param</span> <span class="pl-v">step</span> (optional). Size of the step interval for binning the allele frequency values.</span>
<span class="pl-c">  *</span>
<span class="pl-c">  * <span class="pl-k">@return</span> Returns the distribution of SNPs by allele frequency.</span>
<span class="pl-c">  <span class="pl-c">*/</span></span>
<span class="pl-k">def</span> <span class="pl-en">alleleFrequencyDistribution</span>(<span class="pl-v">step</span>: <span class="pl-k">Double</span> <span class="pl-k">=</span> <span class="pl-en">DEFAULT_ALLELE_FREQUENCY_STEP</span>)</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">zygositySwitchCount</span></pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">allelesSwitchCount</span></pre>
  </div> 
  <h4><a href="https://github.com/tmoerman/vcf-comp#snp-comparison-building-blocks" aria-hidden="true" class="anchor" id="user-content-snp-comparison-building-blocks" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>SNP comparison building blocks</h4> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">countByProjection</span>[<span class="pl-en">P</span>](<span class="pl-v">projection</span>: <span class="pl-en">AnnotatedGenotype</span> <span class="pl-k">=&gt;</span> <span class="pl-en">P</span>)<span class="pl-k">:</span> <span class="pl-en">Iterable</span>[<span class="pl-en">CategoryProjectionCount</span>[<span class="pl-en">P</span>]]</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">countByTraversableProjection</span>[<span class="pl-en">P</span>](<span class="pl-v">projection</span>: <span class="pl-en">AnnotatedGenotype</span> <span class="pl-k">=&gt;</span> <span class="pl-en">Traversable</span>[<span class="pl-en">P</span>])<span class="pl-k">:</span> <span class="pl-en">Iterable</span>[<span class="pl-en">CategoryProjectionCount</span>[<span class="pl-en">P</span>]]</pre>
  </div> 
  <div class="highlight highlight-source-scala">
   <pre><span class="pl-k">def</span> <span class="pl-en">countBySwitch</span>[<span class="pl-en">P</span>](<span class="pl-v">projection</span>: <span class="pl-en">AnnotatedGenotype</span> <span class="pl-k">=&gt;</span> <span class="pl-en">Option</span>[<span class="pl-en">P</span>])<span class="pl-k">:</span> <span class="pl-en">Iterable</span>[<span class="pl-en">CategoryProjectionCount</span>[<span class="pl-k">String</span>]]</pre>
  </div> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#visualizations" aria-hidden="true" class="anchor" id="user-content-visualizations" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Visualizations</h3> 
  <p><code>TODO</code></p> 
  <h2><a href="https://github.com/tmoerman/vcf-comp#how-it-works" aria-hidden="true" class="anchor" id="user-content-how-it-works" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>HOW IT WORKS</h2> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#pimp-my-library" aria-hidden="true" class="anchor" id="user-content-pimp-my-library" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Pimp my library</h3> 
  <p>VCF-comp makes use of a Scala idiom called <a href="http://www.artima.com/weblogs/viewpost.jsp?thread=179766" target="_blank"><em>"Pimp my library"</em></a>, through Scala's implicit conversions.</p> 
  <p><code>TODO</code></p> 
  <h3><a href="https://github.com/tmoerman/vcf-comp#dimplejs" aria-hidden="true" class="anchor" id="user-content-dimplejs" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Dimple.js</h3> 
  <p><code>TODO</code></p> 
  <p><a href="http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf" target="_blank">http://snpeff.sourceforge.net/VCFannotationformat_v1.0.pdf</a></p> 
 </article>
</div>