<div class="announce instapaper_body md" data-path="README.md" id="readme">
 <article class="markdown-body entry-content" itemprop="text">
  <h1><a id="user-content-adam-commands" class="anchor" href="https://github.com/heuermh/adam-commands#adam-commands" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>adam-commands</h1> 
  <p>External commands in Java and Scala for ADAM: Genomic Data System. Apache 2 licensed.</p> 
  <h3><a id="user-content-hacking-adam-commands" class="anchor" href="https://github.com/heuermh/adam-commands#hacking-adam-commands" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Hacking adam-commands</h3> 
  <p>Install</p> 
  <ul> 
   <li>JDK 1.8 or later, <a href="http://openjdk.java.net" target="_blank">http://openjdk.java.net</a></li> 
   <li>Scala 2.10.5 or later, <a href="http://www.scala-lang.org" target="_blank">http://www.scala-lang.org</a></li> 
   <li>Apache Maven 3.2.5 or later, <a href="http://maven.apache.org" target="_blank">http://maven.apache.org</a></li> 
   <li>Apache Spark 2.0.1 or later, <a href="http://spark.apache.org" target="_blank">http://spark.apache.org</a></li> 
   <li>ADAM: Genomic Data System 0.20.0 or later, <a href="https://github.com/bigdatagenomics/adam" target="_blank">https://github.com/bigdatagenomics/adam</a></li> 
  </ul> 
  <p>To build</p> 
  <pre><code>$ mvn install
</code></pre> 
  <h3><a id="user-content-running-adam-commands-using-adam-submit" class="anchor" href="https://github.com/heuermh/adam-commands#running-adam-commands-using-adam-submit" aria-hidden="true" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Running adam-commands using <code>adam-submit</code></h3> 
  <pre><code>$ cp target/adam-commands_2.10-0.20.1-SNAPSHOT.jar $ADAM_DIR
$ cd $ADAM_DIR
</code></pre> 
  <p>To run the external commands in this repository via the ADAM command line, specify <code>ADAM_MAIN</code> and add the adam-commands jar to the classpath with the Spark <code>--jars</code> argument.</p> 
  <p>Note the <code>--</code> argument separator between Spark arguments and ADAM arguments.</p> 
  <p>External commands are now listed in the usage text.</p> 
  <pre><code>$ ADAM_MAIN=com.github.heuermh.adam.commands.ADAMCommandsMain \
  ./bin/adam-submit \
  --jars adam-commands_2.10-0.20.1-SNAPSHOT.jar \
  --

Using ADAM_MAIN=com.github.heuermh.adam.commands.ADAMCommandsMain
Using SPARK_SUBMIT=/usr/local/bin/spark-submit
INFO ADAMMain: ADAM invoked with args:

       e         888~-_          e             e    e
      d8b        888   \        d8b           d8b  d8b
     /Y88b       888    |      /Y88b         d888bdY88b
    /  Y88b      888    |     /  Y88b       / Y88Y Y888b
   /____Y88b     888   /     /____Y88b     /   YY   Y888b
  /      Y88b    888_-~     /      Y88b   /          Y888b

Usage: adam-submit [&lt;spark-args&gt; --] &lt;adam-args&gt;

Choose one of the following commands:

ADAM ACTIONS
             depth : Calculate the depth from a given ADAM file, at each variant in a VCF
       count_kmers : Counts the k-mers/q-mers from a read dataset.
count_contig_kmers : Counts the k-mers/q-mers from a read dataset.
         transform : Convert SAM/BAM to ADAM format and optionally perform read pre-processing transformations
        adam2fastq : Convert BAM to FASTQ files
            plugin : Executes an ADAMPlugin
           flatten : Convert a ADAM format file to a version with a flattened schema, suitable for querying with tools like Impala

CONVERSION OPERATIONS
          vcf2adam : Convert a VCF file to the corresponding ADAM format
         anno2adam : Convert a annotation file (in VCF format) to the corresponding ADAM format
          adam2vcf : Convert an ADAM variant to the VCF ADAM format
        fasta2adam : Converts a text FASTA sequence file into an ADAMNucleotideContig Parquet file which represents assembled sequences.
        adam2fasta : Convert ADAM nucleotide contig fragments to FASTA files
     features2adam : Convert a file with sequence features into corresponding ADAM format
        wigfix2bed : Locally convert a wigFix file to BED format
   fragments2reads : Convert alignment records into fragment records.
   reads2fragments : Convert alignment records into fragment records.

PRINT
             print : Print an ADAM formatted file
       print_genes : Load a GTF file containing gene annotations and print the corresponding gene models
          flagstat : Print statistics on reads in an ADAM file (similar to samtools flagstat)
        print_tags : Prints the values and counts of all tags in a set of records
          listdict : Print the contents of an ADAM sequence dictionary
       allelecount : Calculate Allele frequencies
              view : View certain reads from an alignment-record file.

EXTERNAL COMMANDS
    count_alignments : Counts the alignments in a read dataset.
count_alignments_per_read : Counts the alignments per read in a read dataset.
</code></pre> 
  <p>Run external commands</p> 
  <pre><code>$ ADAM_MAIN=com.github.heuermh.adam.commands.ADAMCommandsMain \
  ./bin/adam-submit \
  --jars adam-commands_2.10-0.20.1-SNAPSHOT.jar \
  -- \
  count_alignments adam-core/src/test/resources/small.sam

Using ADAM_MAIN=com.github.heuermh.adam.commands.ADAMCommandsMain
Using SPARK_SUBMIT=/usr/local/bin/spark-submit
INFO ADAMMain: ADAM invoked with args: "count_alignments" "adam-core/src/test/resources/small.sam"
...
(1,20)


$ ADAM_MAIN=com.github.heuermh.adam.commands.ADAMCommandsMain \
  ./bin/adam-submit \
  --jars adam-commands_2.10-0.20.1-SNAPSHOT.jar \
  -- \
  count_alignments_per_read adam-core/src/test/resources/small.sam

Using ADAM_MAIN=com.github.heuermh.adam.commands.ADAMCommandsMain
Using SPARK_SUBMIT=/usr/local/bin/spark-submit
INFO ADAMMain: ADAM invoked with args: "count_alignments_per_read" "adam-core/src/test/resources/small.sam"
...
(simread:1:237728409:true,1)
(simread:1:195211965:false,1)
(simread:1:163841413:false,1)
(simread:1:231911906:false,1)
(simread:1:26472783:false,1)
(simread:1:165341382:true,1)
(simread:1:240344442:true,1)
(simread:1:50683371:false,1)
(simread:1:240997787:true,1)
(simread:1:14397233:false,1)
(simread:1:207027738:true,1)
(simread:1:20101800:true,1)
(simread:1:5469106:true,1)
(simread:1:186794283:true,1)
(simread:1:189606653:true,1)
(simread:1:101556378:false,1)
(simread:1:37577445:false,1)
(simread:1:89554252:false,1)
(simread:1:153978724:false,1)
(simread:1:169801933:true,1)
</code></pre> 
 </article>
</div>