<div class="announce instapaper_body md" data-path="README.md" id="readme">
 <article class="markdown-body entry-content" itemprop="text">
  <h1><a href="https://github.com/hammerlab/hadoop-bam#hadoop-bam" aria-hidden="true" class="anchor" id="user-content-hadoop-bam" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Hadoop-BAM</h1> 
  <p>Hadoop-BAM: a library for the manipulation of files in common bioinformatics formats using the Hadoop MapReduce framework.</p> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#build-status" aria-hidden="true" class="anchor" id="user-content-build-status" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Build status</h2> 
  <p><a href="https://travis-ci.org/HadoopGenomics/Hadoop-BAM" target="_blank"><img src="https://camo.githubusercontent.com/725c2914f8db4d88f8686ae29ff7286a1c2b10a5/68747470733a2f2f7472617669732d63692e6f72672f4861646f6f7047656e6f6d6963732f4861646f6f702d42414d2e7376673f6272616e63683d6d6173746572" alt="Build Status" data-canonical-src="https://travis-ci.org/HadoopGenomics/Hadoop-BAM.svg?branch=master" style="max-width:100%;"></a> <a href="https://coveralls.io/github/HadoopGenomics/Hadoop-BAM?branch=master" target="_blank"><img src="https://camo.githubusercontent.com/0c561111507f4045595a3dbb623e6112d224afee/68747470733a2f2f636f766572616c6c732e696f2f7265706f732f6769746875622f4861646f6f7047656e6f6d6963732f4861646f6f702d42414d2f62616467652e7376673f6272616e63683d6d6173746572" alt="Coverage Status" data-canonical-src="https://coveralls.io/repos/github/HadoopGenomics/Hadoop-BAM/badge.svg?branch=master" style="max-width:100%;"></a></p> 
  <p>The <a href="http://samtools.github.io/hts-specs/" target="_blank">file formats</a> currently supported are:</p> 
  <ul> 
   <li>BAM (Binary Alignment/Map)</li> 
   <li>SAM (Sequence Alignment/Map)</li> 
   <li>CRAM</li> 
   <li>FASTQ</li> 
   <li>FASTA (input only)</li> 
   <li>QSEQ</li> 
   <li>VCF (Variant Call Format)</li> 
   <li>BCF (Binary VCF) (output is always BGZF-compressed)</li> 
  </ul> 
  <p>For a longer high-level description of Hadoop-BAM, refer to the article "Hadoop-BAM: directly manipulating next generation sequencing data in the cloud" in Bioinformatics Volume 28 Issue 6 pp. 876-877, also available online at: <a href="http://dx.doi.org/10.1093/bioinformatics/bts054" target="_blank">http://dx.doi.org/10.1093/bioinformatics/bts054</a></p> 
  <p>If you are interested in using Apache Pig (<a href="http://pig.apache.org/" target="_blank">http://pig.apache.org/</a>) with Hadoop-BAM, refer to SeqPig at: <a href="http://seqpig.sourceforge.net/" target="_blank">http://seqpig.sourceforge.net/</a></p> 
  <p>Note that the library part of Hadoop-BAM is primarily intended for developers with experience in using Hadoop. SeqPig is a more versatile and higher-level interface to the file formats supported by Hadoop-BAM. In addition, <a href="http://bdgenomics.org/" target="_blank">ADAM</a> and <a href="https://github.com/broadinstitute/gatk" target="_blank">GATK version 4</a> both use Hadoop-BAM and offer high-level command-line bioinformatics tools that run on Spark clusters.</p> 
  <p>For examples of how to use Hadoop-BAM as a library to read data files in Hadoop see the <code>examples/</code> directory.</p> 
  <hr> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#dependencies" aria-hidden="true" class="anchor" id="user-content-dependencies" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Dependencies</h2> 
  <p>Hadoop. Tested with 2.2.0 and later. Version 4.2.0 of Cloudera's distribution, CDH, has also been tested. Use other versions at your own risk. You can change the version of Hadoop linked against by modifying the corresponding parameter in the pom.xml build file.</p> 
  <p>HTSJDK (formerly Picard SAM-JDK) Version 2.3.0 is required. Later versions may also work; Hadoop-BAM currently builds against 2.9.1.</p> 
  <p>Availability:</p> 
  <ul> 
   <li>Hadoop <a href="http://hadoop.apache.org/" target="_blank">http://hadoop.apache.org/</a></li> 
   <li>HTSJDK/Picard <a href="https://github.com/samtools/htsjdk" target="_blank">https://github.com/samtools/htsjdk</a></li> 
  </ul> 
  <p>Note that starting from version 7.4.0 Hadoop-BAM requires Java 8.</p> 
  <hr> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#installation" aria-hidden="true" class="anchor" id="user-content-installation" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Installation</h2> 
  <p>If you're using Hadoop 2.2.0, a precompiled "hadoop-bam-X.Y.Z.jar" is available that you can use. Otherwise, you'll have to build Hadoop-BAM yourself by by using Maven (version 3.0.4 at least) and following the instructions below.</p> 
  <h3><a href="https://github.com/hammerlab/hadoop-bam#hadoop-version" aria-hidden="true" class="anchor" id="user-content-hadoop-version" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Hadoop version</h3> 
  <p>You must set the &lt;hadoop.version&gt; tag in pom.xml appropriately for the version of hadoop you're using. Run "hadoop version" and copy the string from the output:</p> 
  <pre><code>[luca@vm Hadoop-BAM]# hadoop version 2&gt;/dev/null | grep Hadoop
Hadoop 2.0.0-cdh4.6.0
</code></pre> 
  <p>In the output above, the version string is "2.0.0-cdh4.6.0".</p> 
  <h3><a href="https://github.com/hammerlab/hadoop-bam#build" aria-hidden="true" class="anchor" id="user-content-build" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Build</h3> 
  <p>Build Hadoop-BAM with the following command:</p> 
  <pre><code>mvn clean package -DskipTests
</code></pre> 
  <p>It will create two files:</p> 
  <pre><code>target/hadoop-bam-X.Y.Z-SNAPSHOT.jar
target/hadoop-bam-X.Y.Z-SNAPSHOT-jar-with-dependencies.jar
</code></pre> 
  <p>The former contains only Hadoop-BAM whereas the latter one also contains all dependencies and can be run directly via</p> 
  <pre><code>hadoop jar target/hadoop-bam-X.Y.Z-SNAPSHOT-jar-with-dependencies.jar
</code></pre> 
  <p>Javadoc documentation is generated automatically and can then be found in the <code>target/apidocs</code> subdirectory.</p> 
  <p>Finally, unit tests can be run via:</p> 
  <pre><code>mvn test
</code></pre> 
  <hr> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#library-usage" aria-hidden="true" class="anchor" id="user-content-library-usage" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Library usage</h2> 
  <p>Hadoop-BAM provides the standard set of Hadoop file format classes for the file formats it supports: a FileInputFormat and one or more RecordReaders for input, and a FileOutputFormat and one or more RecordWriters for output. These are summarized in the table below.</p> 
  <table> 
   <thead> 
    <tr> 
     <th>File Format</th> 
     <th>InputFormat</th> 
     <th>OutputFormat</th> 
    </tr> 
   </thead> 
   <tbody> 
    <tr> 
     <td>BAM</td> 
     <td><code>BAMInputFormat</code></td> 
     <td><code>KeyIgnoringBAMOutputFormat</code></td> 
    </tr> 
    <tr> 
     <td>SAM</td> 
     <td><code>SAMInputFormat</code></td> 
     <td><code>KeyIgnoringAnySAMOutputFormat</code></td> 
    </tr> 
    <tr> 
     <td>CRAM</td> 
     <td><code>CRAMInputFormat</code></td> 
     <td><code>KeyIgnoringCRAMOutputFormat</code></td> 
    </tr> 
    <tr> 
     <td>BAM, SAM, or CRAM</td> 
     <td><code>AnySAMInputFormat</code></td> 
     <td><code>KeyIgnoringAnySAMOutputFormat</code></td> 
    </tr> 
    <tr> 
     <td>FASTQ</td> 
     <td><code>FastqInputFormat</code></td> 
     <td><code>FastqOutputFormat</code></td> 
    </tr> 
    <tr> 
     <td>FASTA</td> 
     <td><code>FastaInputFormat</code></td> 
     <td>N/A</td> 
    </tr> 
    <tr> 
     <td>QSEQ</td> 
     <td><code>QseqInputFormat</code></td> 
     <td><code>QseqOutputFormat</code></td> 
    </tr> 
    <tr> 
     <td>VCF or BCF</td> 
     <td><code>VCFInputFormat</code></td> 
     <td><code>KeyIgnoringVCFOutputFormat</code></td> 
    </tr>
   </tbody>
  </table> 
  <p><code>AnySAMInputFormat</code> detects the format (BAM, SAM, or CRAM) by file extension, then by looking at the first few bytes of the file (if file extension detection is disabled or is inconclusive). <code>VCFInputFormat</code> works in a similar way for VCF and BCF files.</p> 
  <p>The output formats all discard the key and only use the value field when writing the output file. Some of the output formats indicate this explictly through the <code>KeyIgnoring</code> prefix in their name, but <code>FastqOutputFormat</code> and <code>QseqOutputFormat</code> actually ignore the key too.</p> 
  <p>The abstract base classes <code>BAMOutputFormat</code>, <code>CRAMOutputFormat</code>, <code>AnySAMOutputFormat</code>, and <code>VCFOutputFormat</code> cannot be used directly, but can be subclassed in order to add custom logic (to provide <code>ValueIgnoring</code> versions, for example).</p> 
  <p>When using <code>KeyIgnoringAnySAMOutputFormat</code>, the format of the files written (BAM, SAM, or CRAM) must be specified by setting the property <code>hadoopbam.anysam.output-format</code>. Similarly, set the property <code>hadoopbam.vcf.output-format</code> in order to specify which file format <code>KeyIgnoringVCFOutputFormat</code> will use (VCF or BCF).</p> 
  <p>The properties that can be set on input and output formats are summarized in the table below:</p> 
  <table> 
   <thead> 
    <tr> 
     <th>Format</th> 
     <th>Property</th> 
     <th>Default</th> 
     <th>Description</th> 
    </tr> 
   </thead> 
   <tbody> 
    <tr> 
     <td><code>AnySAMInputFormat</code></td> 
     <td><code>hadoopbam.anysam.trust-exts</code></td> 
     <td><code>true</code></td> 
     <td>Whether to detect the file format (BAM, SAM, or CRAM) by file extension. If <code>false</code>, use the file contents to detect the format.</td> 
    </tr> 
    <tr> 
     <td><code>KeyIgnoringAnySAMOutputFormat</code></td> 
     <td><code>hadoopbam.anysam.output-format</code></td> 
     <td></td> 
     <td>(Required.) The file format to use when writing BAM, SAM, or CRAM files. Should be one of <code>BAM</code>, <code>SAM</code>, or <code>CRAM</code>.</td> 
    </tr> 
    <tr> 
     <td></td> 
     <td><code>hadoopbam.anysam.write-header</code></td> 
     <td><code>true</code></td> 
     <td>Whether to write the SAM header in each output file part. If <code>true</code>, call <code>setSAMHeader()</code> or <code>readSAMHeaderFrom()</code> to set the desired header.</td> 
    </tr> 
    <tr> 
     <td><code>BAMInputFormat</code></td> 
     <td><code>hadoopbam.bam.intervals</code></td> 
     <td></td> 
     <td>Only include reads that match the specified intervals. BAM files must be indexed in this case. Intervals are comma-separated and follow the same syntax as the <code>-L</code> option in SAMtools. E.g. <code>chr1:1-20000,chr2:12000-20000</code>.</td> 
    </tr> 
    <tr> 
     <td><code>KeyIgnoringBAMOutputFormat</code></td> 
     <td><code>hadoopbam.bam.write-splitting-bai</code></td> 
     <td><code>false</code></td> 
     <td>If <code>true</code>, write <em>.splitting-bai</em> files for every BAM file.</td> 
    </tr> 
    <tr> 
     <td><code>CRAMInputFormat</code></td> 
     <td><code>hadoopbam.cram.reference-source-path</code></td> 
     <td></td> 
     <td>(Required.) The path to the reference. May be an <code>hdfs://</code> path.</td> 
    </tr> 
    <tr> 
     <td><code>FastqInputFormat</code></td> 
     <td><code>hbam.fastq-input.base-quality-encoding</code></td> 
     <td><code>sanger</code></td> 
     <td>The encoding used for base qualities. One of <code>sanger</code> or <code>illumina</code>.</td> 
    </tr> 
    <tr> 
     <td></td> 
     <td><code>hbam.fastq-input.filter-failed-qc</code></td> 
     <td><code>false</code></td> 
     <td>If <code>true</code>, filter out reads that didn't pass quality checks.</td> 
    </tr> 
    <tr> 
     <td><code>QseqInputFormat</code></td> 
     <td><code>hbam.qseq-input.base-quality-encoding</code></td> 
     <td><code>illumina</code></td> 
     <td>The encoding used for base qualities. One of <code>sanger</code> or <code>illumina</code>.</td> 
    </tr> 
    <tr> 
     <td></td> 
     <td><code>hbam.qseq-input.filter-failed-qc</code></td> 
     <td><code>false</code></td> 
     <td>If <code>true</code>, filter out reads that didn't pass quality checks.</td> 
    </tr> 
    <tr> 
     <td><code>VCFInputFormat</code></td> 
     <td><code>hadoopbam.vcf.trust-exts</code></td> 
     <td><code>false</code></td> 
     <td>Whether to detect the file format (VCF or BCF) by file extension. If <code>false</code>, use the file contents to detect the format.</td> 
    </tr> 
    <tr> 
     <td><code>KeyIgnoringVCFOutputFormat</code></td> 
     <td><code>hadoopbam.vcf.output-format</code></td> 
     <td></td> 
     <td>(Required.) The file format to use when writing VCF or BCF files. Should be one of <code>VCF</code> or <code>BCF</code>.</td> 
    </tr> 
    <tr> 
     <td></td> 
     <td><code>hadoopbam.vcf.write-header</code></td> 
     <td><code>true</code></td> 
     <td>Whether to write the VCF header in each output file part. If <code>true</code>, call <code>setHeader()</code> or <code>readHeaderFrom()</code> to set the desired header.</td> 
    </tr>
   </tbody>
  </table> 
  <p>Note that Hadoop-BAM is based around the newer Hadoop API introduced in the 0.20 Hadoop releases instead of the older, deprecated API.</p> 
  <p>For examples of how to link to Hadoop-BAM in your own Maven project see the <code>examples/</code> folder. There are examples for reading and writing BAM as well as VCF files.</p> 
  <p>When using Hadoop-BAM as a library in your program, remember to have hadoop-bam-X.Y.Z.jar as well as the HTSJDK .jars (including the Commons JEXL .jar) in your CLASSPATH and HADOOP_CLASSPATH; alternatively, use the *-jar-with-dependencies.jar which contains already all dependencies.</p> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#linking-against-hadoop-bam" aria-hidden="true" class="anchor" id="user-content-linking-against-hadoop-bam" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Linking against Hadoop-BAM</h2> 
  <p>If your Maven project relies on Hadoop-BAM the easiest way to link against it is by relying on the OSS Sonatype repository:</p> 
  <pre><code>&lt;project&gt;
...
    &lt;dependencies&gt;
        &lt;dependency&gt;
            &lt;groupId&gt;org.seqdoop&lt;/groupId&gt;
            &lt;artifactId&gt;hadoop-bam&lt;/artifactId&gt;
            &lt;version&gt;7.8.0&lt;/version&gt;
        &lt;/dependency&gt;
        ...
    &lt;/dependencies&gt;
    ...
&lt;/project&gt;
</code></pre> 
  <hr> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#running-under-hadoop" aria-hidden="true" class="anchor" id="user-content-running-under-hadoop" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>Running under Hadoop</h2> 
  <p>To use Hadoop-BAM under Hadoop, the easiest method is to use the jar that comes packaged with all dependencies via</p> 
  <pre><code>hadoop jar hadoop-bam-X.Y.Z-jar-with-dependencies.jar
</code></pre> 
  <p>Alternatively, you can use the "-libjars" command line argument when running Hadoop-BAM to provide different versions of dependencies as follows:</p> 
  <pre><code>hadoop jar hadoop-bam-X.Y.Z.jar \
  -libjars htsjdk-2.3.0.jar,commons-jexl-2.1.1.jar
</code></pre> 
  <p>Finally, all jar files can also be added to HADOOP_CLASSPATH in the Hadoop configuration's hadoop-env.sh.</p> 
  <h2><a href="https://github.com/hammerlab/hadoop-bam#file-paths-under-hadoop" aria-hidden="true" class="anchor" id="user-content-file-paths-under-hadoop" target="_blank">
    <svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewbox="0 0 16 16" width="16">
     <path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path>
    </svg></a>File paths under Hadoop</h2> 
  <p>When running under Hadoop, keep in mind that file paths refer to the distributed file system, HDFS. To explicitly access a local file, instead of using the plain path such as "/foo/bar", you must use a file: URI, such as "file:/foo/bar". Note that paths in file: URIs must be absolute.</p> 
 </article>
</div>